### PR TITLE
Handle null ranges in changes in SourceWriter#applyChanges

### LINKED ISSUE #112: GroovyTextDocumentService#didChange should replace whole content of file if Range is null
- If range for a particular change to a text document is null, assert it is the only change.

### PR DESCRIPTION
**Pull request by natgabb**
Closes #112

### CODE REVIEW

File: groovy-language-server/src/main/java/com/palantir/ls/groovy/util/SourceWriter.java [hunk 1]
```diff
@@ -69,6 +69,19 @@ public Path getDestination() {
...
      */
     public synchronized void applyChanges(List<TextDocumentContentChangeEvent> contentChanges) throws IOException {
+        // Check if any of the ranges are null
+        for (TextDocumentContentChangeEvent change : contentChanges) {
+            if (change.getRange() == null) {
+                checkArgument(contentChanges.size() == 1,
+                        String.format(
+                                "Cannot have many changes when a change contains a null range which means it replaces "
+                                        + "the whole contents of the file: %s",
+                                change.toString()));
```
Comment (jiahuijiang) [inline groovy-language-server/src/main/java/com/palantir/ls/groovy/util/SourceWriter.java:77]: hmmm for the error message feel it should also provide the file uri

File: groovy-language-server/src/main/java/com/palantir/ls/groovy/util/SourceWriter.java [hunk 2]
```diff
@@ -69,6 +69,19 @@ public Path getDestination() {
      * Applies the given changes to the destination file. Does not handle intersecting ranges in the changes.
      */
     public synchronized void applyChanges(List<TextDocumentContentChangeEvent> contentChanges) throws IOException {
+        // Check if any of the ranges are null
+        for (TextDocumentContentChangeEvent change : contentChanges) {
+            if (change.getRange() == null) {
+                checkArgument(contentChanges.size() == 1,
+                        String.format(
+                                "Cannot have many changes when a change contains a null range which means it replaces "
```
Comment (jiahuijiang) [inline groovy-language-server/src/main/java/com/palantir/ls/groovy/util/SourceWriter.java]: something like `Cannot handle more than one change in one single request when some of them have a null range`? Probably don't need to explain what null range means

File: groovy-language-server/src/main/java/com/palantir/ls/groovy/util/SourceWriter.java (hunk 1 above)
Comment (natgabb) [reply groovy-language-server/src/main/java/com/palantir/ls/groovy/util/SourceWriter.java:77]: It doesn't really matter which file it is though, the change is what is invalid. We also don't print out the URI below.
The thing is we won't know what the workspace root is in here (it's kind of irrelevant to the SourceWriter) and I wouldn't want to leak the absolute path.

Comment (jiahuijiang) [reply groovy-language-server/src/main/java/com/palantir/ls/groovy/util/SourceWriter.java:77]: 👍 makes sense

Comment (natgabb) [reply groovy-language-server/src/main/java/com/palantir/ls/groovy/util/SourceWriter.java:77]: I just realized we're leaking absolute paths elsewhere :( we should fix that #114

File: groovy-language-server/src/main/java/com/palantir/ls/groovy/util/SourceWriter.java (hunk 2 above)
Comment (natgabb) [reply groovy-language-server/src/main/java/com/palantir/ls/groovy/util/SourceWriter.java]: How about: `Cannot handle more than one change when a null range exists:`

**Pull request merged**